http/client/src/main/java/org/wildfly/security/http/client/ElytronHttpClient.java

```diff
@@ -0,0 +1,139 @@
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Elytron client for HTTP authentication
+ *
+ * @author <a href="mailto:kekumar@redhat.com">Keshav Kumar</a>
+ */
+public class ElytronHttpClient {
+
+    private HttpClient httpClient;
+
+    public ElytronHttpClient() {
+        this.httpClient = HttpClient.newHttpClient();
+    }
+
+    private HttpResponse getResponse(HttpRequest request) throws IOException, InterruptedException {
+        return httpClient.send(request, HttpResponse.BodyHandlers.ofString());
+    }
+
+    private void addSSLContextToHttpClient(URI uri) throws MalformedURLException {
+        String protocol = new URL(uri.toString()).getProtocol();
+
+        if (protocol.equalsIgnoreCase("https")) {
+            SSLContext sslContext = ElytronHttpClientCredentialUtils.getSSLContext(uri);
+            httpClient = HttpClient.newBuilder().sslContext(sslContext).build();
+        }
+    }
+
+    /**
+     * Used to connect to the secured uri and return the response based on that.
+     */
+    public HttpResponse connect(String uri, String method, String body, Map<String, String> headers) throws IOException, InterruptedException, URISyntaxException {
+        URI uriPath = new URI(uri);
+        addSSLContextToHttpClient(uriPath);
+
+        HttpRequest request = ElytronHttpClientRequestBuilder.buildRequest(uriPath, method, body, headers);
+        HttpResponse response = getResponse(request);
+
+        if (response.statusCode() == OK) {
+            return response;
+        }
+
+        String authHeader = getAuthHeader(response);
+
+        if (authHeader == null) {
+            throw new ElytronHttpClientException(ElytronMessages.log.responseHeaderExtractionFailed());
+        }
+        String[] authChallenges = authHeader.split(",");
+        HttpRequest authRequest = null;
+
+        String challenge = authChallenges[0];
+
+        String authType = getAuthType(challenge);
+
+        switch (authType) {
+            case "basic":
+                authRequest = ElytronHttpClientBasicAuthMechanism.evaluateMechanism(uriPath, method, body, headers);
+                break;
+            case "digest":
+                authRequest = ElytronHttpClientDigestAuthMechanism.evaluateMechanism(uriPath, authHeader, method, body, headers);
+                break;
+            case "bearer":
+                authRequest = ElytronHttpClientBearerAuthMechanism.evaluateMechanism(uriPath, method, body, headers);
+                break;
+            default:
+                authRequest = ElytronHttpClientRequestBuilder.buildRequest(uriPath, method, body, headers);
+        }
+
+        if (authRequest != null) {
+            response = getResponse(authRequest);
+            return response;
+        }
+
+        // If none of the authentication mechanisms succeeded, fallback to the initial request
+        response = getResponse(request);
```

Comment:
@keshav-725  Just wondering if getResponse(request) has to be called the second time here?